src/lib/optimizer/strategy/constant_calculation_rule.cpp

```diff
@@ -47,54 +62,107 @@ bool ConstantCalculationRule::apply_to(const std::shared_ptr<AbstractLQPNode>& n
-                                                             const AllTypeVariant& value) {
-  auto parent_tree_changed = false;
-  for (auto parent : node->parents()) {
-    if (parent->type() != LQPNodeType::Predicate) {
-      parent_tree_changed |= _replace_expression_in_parents(parent, expression_column, value);
-      continue;
-    }
+bool ConstantCalculationRule::_replace_column_references_in_tree(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    const std::map<LQPColumnReference, AllTypeVariant>& column_reference_to_value_map) {
+  auto tree_changed = false;
 
-    auto predicate_node = std::static_pointer_cast<PredicateNode>(parent);
+  if (node->left_child()) {
+    tree_changed |= _replace_column_references_in_tree(node->left_child(), column_reference_to_value_map);
+  }
+  if (node->right_child()) {
+    tree_changed |= _replace_column_references_in_tree(node->right_child(), column_reference_to_value_map);
+  }
+
+  if (node->type() == LQPNodeType::Predicate) {
+    auto predicate_node = std::static_pointer_cast<PredicateNode>(node);
 
     // We look for a PredicateNode which has an LQPColumnReference as value,
     // referring to the column which contains the Expression we resolved before.
-    if (is_lqp_column_reference(predicate_node->value()) &&
-        boost::get<LQPColumnReference>(predicate_node->value()) == expression_column) {
-      // We replace the LQPColumnReference with the actual result of the Expression it was referring to.
-      auto new_predicate_node = std::make_shared<PredicateNode>(predicate_node->column_reference(),
-                                                                predicate_node->predicate_condition(), value);
-      predicate_node->replace_with(new_predicate_node);
-      parent_tree_changed = true;
+    if (is_lqp_column_reference(predicate_node->value())) {
+      const auto column_reference = boost::get<LQPColumnReference>(predicate_node->value());
+      if (column_reference_to_value_map.find(column_reference) != column_reference_to_value_map.end()) {
+        // We replace the LQPColumnReference with the actual result of the Expression it was referring to.
+        auto new_predicate_node =
+            std::make_shared<PredicateNode>(predicate_node->column_reference(), predicate_node->predicate_condition(),
+                                            column_reference_to_value_map.at(column_reference));
+        predicate_node->replace_with(new_predicate_node);
+        tree_changed = true;
+      }
+    }
+  } else if (node->type() == LQPNodeType::Projection) {
+    auto projection_node = std::static_pointer_cast<ProjectionNode>(node);
+    auto new_column_expressions = std::vector<std::shared_ptr<LQPExpression>>();
+    auto expression_replaced = false;
+
+    for (const auto expression : projection_node->column_expressions()) {
+      if (expression->type() == ExpressionType::Column &&
```

Comment:
What about nested Expressions? (e.g., a > 5)
They wouldn't get their LQPColumnReferences replaced with values, leading to invalid LQPColumnReferences in the LQP.

Reply:
You're right, I added a method to recursively replace the column references in a (potentially nested) expression.